### PR TITLE
Manage plugin factory plugins with unique_ptr in Alignment

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR is a preparatory step to change PluginFactory to return `std::unique_ptr`.

In addition use range-for, and more `unique_ptr` in AlignmentProducerBase.

#### PR validation:

Code compiles.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: Alignment/ReferenceTrajectories/plugins/CombinedTrajectoryFactory.cc [hunk 1]
```diff
@@ -38,7 +38,15 @@ class CombinedTrajectoryFactory : public TrajectoryFactoryBase {
...
-  std::vector<TrajectoryFactoryBase *> theFactories;
+  CombinedTrajectoryFactory(const CombinedTrajectoryFactory &other)
+      : TrajectoryFactoryBase(other), theUseAllFactories{other.theUseAllFactories} {
+    theFactories.reserve(other.theFactories.size());
+    for (const auto &f : other.theFactories) {
+      theFactories.emplace_back(f->clone());
+    }
+  }
+
+  std::vector<std::unique_ptr<TrajectoryFactoryBase>> theFactories;
```
Comment (makortel) [inline Alignment/ReferenceTrajectories/plugins/CombinedTrajectoryFactory.cc]: This vector could be also done with `std::shared_ptr`, which would avoid defining the custom copy constructor (with the cloning), and would keep the `TrajectoryFactoryBase` objects shared between the copies of `CombinedTrajectoryFactory`.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by pohsun** [comment]
+1

File: Alignment/ReferenceTrajectories/plugins/CombinedTrajectoryFactory.cc (hunk 1 above)
Comment (fabiocos) [reply Alignment/ReferenceTrajectories/plugins/CombinedTrajectoryFactory.cc]: @makortel are you satisfied with this PR in its present status or are you planning  further update?

Comment (makortel) [reply Alignment/ReferenceTrajectories/plugins/CombinedTrajectoryFactory.cc]: I'm satisfied, just wanted to say that it is possible to also share the `TrajectoryFactoryBase` objects if that is needed. So I'm not planning updates.

**Comment by kpedro88** [comment]
+1

**Pull request merged**